services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/RequestBatchManager.java

```diff
@@ -0,0 +1,198 @@
+public abstract class RequestBatchManager<RequestT, ResponseT, BatchResponseT> {
+    private final int maxBatchItems;
+    private final Duration maxBatchOpenInMs;
+    private final BatchingMap<RequestT, ResponseT> requestsAndResponsesMaps;
+    private final BatchAndSend<RequestT, BatchResponseT> batchFunction;
+    private final BatchResponseMapper<BatchResponseT, ResponseT> responseMapper;
+    private final BatchKeyMapper<RequestT> batchKeyMapper;
+    private final ScheduledExecutorService scheduledExecutor;
+
+    protected RequestBatchManager(DefaultBuilder<RequestT, ResponseT, BatchResponseT, ?> builder) {
+        BatchConfiguration batchConfiguration = new BatchConfiguration(builder.overrideConfiguration);
+        this.requestsAndResponsesMaps = new BatchingMap<>(batchConfiguration.maxBatchKeys(),
+                                                          batchConfiguration.maxBufferSize(),
+                                                          RequestBatchBuffer::new);
+        this.maxBatchItems = batchConfiguration.maxBatchItems();
+        this.maxBatchOpenInMs = batchConfiguration.maxBatchOpenInMs();
+        this.batchFunction = Validate.notNull(builder.batchFunction, "Null batchFunction");
+        this.responseMapper = Validate.notNull(builder.responseMapper, "Null responseMapper");
+        this.batchKeyMapper = Validate.notNull(builder.batchKeyMapper, "Null batchKeyMapper");
+        this.scheduledExecutor = Validate.notNull(builder.scheduledExecutor, "Null scheduledExecutor");
+    }
+
+    public CompletableFuture<ResponseT> batchRequest(RequestT request) {
+        CompletableFuture<ResponseT> response = new CompletableFuture<>();
+        try {
+            String batchKey = batchKeyMapper.getBatchKey(request);
+            requestsAndResponsesMaps.put(batchKey,
+                                         () -> scheduleBufferFlush(batchKey, maxBatchOpenInMs.toMillis(), scheduledExecutor),
+                                         request,
+                                         response);
+            flushBufferIfNeeded(batchKey);
+        } catch (Exception e) {
+            response.completeExceptionally(e);
+        }
+        return response;
+    }
+
+    private void flushBufferIfNeeded(String batchKey) {
+        Map<String, BatchingExecutionContext<RequestT, ResponseT>> flushableRequests =
+            requestsAndResponsesMaps.flushableRequests(batchKey, maxBatchItems);
+        if (!flushableRequests.isEmpty()) {
+            manualFlushBuffer(batchKey, flushableRequests);
+        }
+    }
+
+    private void manualFlushBuffer(String batchKey,
+                                   Map<String, BatchingExecutionContext<RequestT, ResponseT>> flushableRequests) {
+        requestsAndResponsesMaps.cancelScheduledFlush(batchKey);
+        flushBuffer(batchKey, flushableRequests);
+        requestsAndResponsesMaps.putScheduledFlush(batchKey, scheduleBufferFlush(batchKey, maxBatchOpenInMs.toMillis(),
```

Comment:
What happens if the flush can't be cancelled because it's already running?

Reply:
Currently it would just wait until response is recieved and relied on user to cancel it or complete it , but now I have fixed it to explicitly cancel them whenever BatchManager is closed